File: demo-app/app/src/javaApp/java/com/launchkey/android/authenticator/demo/ui/fragment/SecurityInfoFragment.java
package com.launchkey.android.authenticator.demo.ui.fragment;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.launchkey.android.authenticator.demo.R;
import com.launchkey.android.authenticator.demo.databinding.DemoFragmentSecurityInfoBinding;
import com.launchkey.android.authenticator.demo.ui.fragment.security.SecurityFactor;
import com.launchkey.android.authenticator.demo.ui.fragment.security.SecurityService;

import java.util.List;

public class SecurityInfoFragment extends BaseDemoFragment<DemoFragmentSecurityInfoBinding> implements SecurityService.SecurityStatusListener {
    @NonNull
    private final SecurityService securityService = SecurityService.getInstance();

    public SecurityInfoFragment() {
        super(R.layout.demo_fragment_security_info);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        binding = DemoFragmentSecurityInfoBinding.bind(view);
    }

    @Override
    public void onResume() {
        super.onResume();
        securityService.getStatus(this);
    }

    @Override
    public void onSecurityStatusUpdate(List<SecurityFactor> list) {
        showSecurityInformation(list);
    }

    private void showSecurityInformation(List<SecurityFactor> list) {

        if (getActivity() == null || getActivity().isFinishing()) {
            return;
        }

        StringBuilder sb = new StringBuilder();

        for (SecurityFactor f : list) {

            sb.append("\n");

            if (f != null) {
                sb.append(getString(R.string.demo_activity_list_feature_security_info_line,
                        getFactorName(f), getFactorCategory(f),
                        f.isActive() ? getString(R.string.demo_generic_true) : getString(R.string.demo_generic_false)));
            } else {
                sb.append(getString(R.string.demo_activity_list_feature_security_info_null_factor));
            }
        }

        String message = sb.toString();

        if (message.trim().isEmpty()) {
            message = getString(R.string.demo_activity_list_feature_security_info_no_message);
        }

        binding.demoFragmentSecurityinfoText.setText(message);
    }

    private String getFactorName(SecurityFactor f) {
        switch (f.getAuthMethod()) {
            case PIN_CODE:
                return getString(R.string.demo_activity_list_feature_security_info_pin_code);
            case CIRCLE_CODE:
                return getString(R.string.demo_activity_list_feature_security_info_circle_code);
            case WEARABLES:
                return getString(R.string.demo_activity_list_feature_security_info_bluetooth_devices);
            case LOCATIONS:
                return getString(R.string.demo_activity_list_feature_security_info_geofencing);
            case BIOMETRIC:
                return getString(R.string.demo_activity_list_feature_security_info_fingerprint);
            default:
                return getString(R.string.demo_activity_list_feature_security_info_none);
        }
    }

    private String getFactorCategory(SecurityFactor f) {
        switch (f.getType()) {
            case KNOWLEDGE:
                return getString(R.string.demo_activity_list_feature_security_info_knowledge);
            case INHERENCE:
                return getString(R.string.demo_activity_list_feature_security_info_inherence);
            case POSSESSION:
                return getString(R.string.demo_activity_list_feature_security_info_possession);
            default:
                return getString(R.string.demo_activity_list_feature_security_info_none);
        }
    }
}

File: demo-app/app/src/kotlinApp/java/com/launchkey/android/authenticator/demo/ui/fragment/SecurityInfoFragment.kt
package com.launchkey.android.authenticator.demo.ui.fragment

import android.os.Bundle
import android.view.View
import com.launchkey.android.authenticator.demo.R
import com.launchkey.android.authenticator.demo.databinding.DemoFragmentSecurityInfoBinding
import com.launchkey.android.authenticator.demo.ui.fragment.security.SecurityFactor
import com.launchkey.android.authenticator.demo.ui.fragment.security.SecurityService
import com.launchkey.android.authenticator.demo.ui.fragment.security.SecurityService.SecurityStatusListener
import com.launchkey.android.authenticator.sdk.core.auth_method_management.AuthMethod
import com.launchkey.android.authenticator.sdk.core.auth_method_management.AuthMethodType

class SecurityInfoFragment : BaseDemoFragment<DemoFragmentSecurityInfoBinding>(R.layout.demo_fragment_security_info), SecurityStatusListener {
    private val securityService: SecurityService = SecurityService.instance
    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        binding = DemoFragmentSecurityInfoBinding.bind(view)
    }

    override fun onResume() {
        super.onResume()
        securityService.getStatus(this)
    }

    override fun onSecurityStatusUpdate(list: List<SecurityFactor>) {
        showSecurityInformation(list)
    }

    private fun showSecurityInformation(list: List<SecurityFactor>) {
        if (activity == null || activity!!.isFinishing) {
            return
        }
        val sb = StringBuilder()
        for (f in list) {
            sb.append("\n")
            if (f != null) {
                sb.append(getString(R.string.demo_activity_list_feature_security_info_line,
                        getFactorName(f), getFactorCategory(f),
                        if (f.isActive) getString(R.string.demo_generic_true) else getString(R.string.demo_generic_false)))
            } else {
                sb.append(getString(R.string.demo_activity_list_feature_security_info_null_factor))
            }
        }
        var message = sb.toString()
        if (message.trim { it <= ' ' }.isEmpty()) {
            message = getString(R.string.demo_activity_list_feature_security_info_no_message)
        }
        binding!!.demoFragmentSecurityinfoText.text = message
    }

    private fun getFactorName(f: SecurityFactor): String {
        return when (f.authMethod) {
            AuthMethod.PIN_CODE -> getString(R.string.demo_activity_list_feature_security_info_pin_code)
            AuthMethod.CIRCLE_CODE -> getString(R.string.demo_activity_list_feature_security_info_circle_code)
            AuthMethod.WEARABLES -> getString(R.string.demo_activity_list_feature_security_info_bluetooth_devices)
            AuthMethod.LOCATIONS -> getString(R.string.demo_activity_list_feature_security_info_geofencing)
            AuthMethod.BIOMETRIC -> getString(R.string.demo_activity_list_feature_security_info_fingerprint)
            else -> getString(R.string.demo_activity_list_feature_security_info_none)
        }
    }

    private fun getFactorCategory(f: SecurityFactor): String {
        return when (f.type) {
            AuthMethodType.KNOWLEDGE -> getString(R.string.demo_activity_list_feature_security_info_knowledge)
            AuthMethodType.INHERENCE -> getString(R.string.demo_activity_list_feature_security_info_inherence)
            AuthMethodType.POSSESSION -> getString(R.string.demo_activity_list_feature_security_info_possession)
            else -> getString(R.string.demo_activity_list_feature_security_info_none)
        }
    }
}